### PR TITLE
Allow for not compressed files, use ujson

### PR DESCRIPTION
**Pull request by maxnoe**

### CODE REVIEW

File: photon_stream/JsonLinesReader.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    Sequentially reads a JSON-Lines file line by line and provides
+    dictionaries for each line. Also supports gzipped files.
+    """
+    def __init__(self, path):
+        self.path = os.path.abspath(path)
+
+        # check for gzip file
+        with open(path, 'rb') as f:
+            marker = f.read(2)
+            gzipped = marker[0] == 31 and marker[1] == 139
```
Comment (kbruegge) [inline photon_stream/JsonLinesReader.py]: Maybe add a link to https://en.wikipedia.org/wiki/Gzip or even https://tools.ietf.org/html/rfc1952#page-5 in the comment.

**Review by kbruegge — APPROVED** [review]

**Comment by dneise** [comment]
This PR drops some features from the `JsonLinesGzipReader`, like this [checking for a comma in the end of a line](https://github.com/fact-project/photon_stream/pull/15/files#diff-c79cf9d656f70bbc208ddaf51098bea9L22)

It also opens all files in read binary mode `rb` and then does this [decode](https://github.com/fact-project/photon_stream/pull/15/files#diff-1af8423b6e5b8106569695f02033ff4fR31) manually. Whereas the previous implementation opened in [text-mode](https://github.com/fact-project/photon_stream/pull/15/files#diff-c79cf9d656f70bbc208ddaf51098bea9L12)

I am not against these changes, but I don't get why they have been made. I think a quick comment regarding this would be in order.

**Comment by maxnoe** [comment]
> This PR drops some features from the JsonLinesGzipReader, like this checking for a comma in the end of a line

No it doesn't:
```python
line = self.file.readline().strip().rstrip(',')
```
See the `rstrip(',')`?

What are the other features, which got dropped?

> It also opens all files in read binary mode rb and then does this decode manually. Whereas the previous implementation opened in text-mode

I changed it back, dunno why I opened it in binary and used `decode`.

File: photon_stream/JsonLinesReader.py (hunk 1 above)
Comment (maxnoe) [reply photon_stream/JsonLinesReader.py]: done

**Comment by dneise** [comment]
Ah I missed the rstrip.
Okay thanks for the clarification.

**Review by dneise — APPROVED** [review]

**Comment by dneise** [comment]
Oh, I just see `__enter__` is missing

**Comment by dneise** [comment]
An I think `__exit__` has this interface?

    def __exit__(self, exc_type, exc_value, traceback):
        pass

not this

    def __exit__(self):
        pass

**Comment by maxnoe** [comment]
I fixed the context managers for JsonLineReader and EventListReader

**Comment by dneise** [comment]
Perfect .. sorry I was too lazy

**Comment by relleums** [comment]
It happens too much here.

I assume we want to use ujson because it is faster. Is there a rough estimate on the performance gain?
Could we introduce ujson independent from the optional gzip usage?
Might it be, that this could be done with a one line pull request? 
```import ujson as json```

I assume there is somebody who needs the optional gzip reading right now, and I am not against it.
However, I would like to know her work-flow to plan future refactoring. In case this is just a 'lets do it because we can' feature, I would not like to have it as it makes the reader more complicated. In my first implementation I also had this feature. Personally I would make the gzip decision not on the magic bytes but on the file ending ```.gz```. I think this could be a separated pull request, too. 

The crazy comma striping in the end of lines is a good catch, I forgot about it. It was needed in the very early stages when I was not able to use fact-tools correctly. I think the comma striping can be removed entirely now. However, this is again a separate pull request.

**Comment by relleums** [comment]
In general, please add unit tests. For my personal work-flow I have a pretty high test coverage. So I can keep my stuff running during refactorings. But untested features (e.g. optional gzip reading) are (we all know this) very likely to break when not covered in tests during the next refactoring.

**Comment by relleums** [comment]
The usage of  ```ujson``` is indeed a one liner plus the another line in ```setup.py```. However, in the current unit tests (which only represent my narrow minded work-flow) is no speed up at all. I am afraid the DBSCAN clustering is already the bottleneck. 
Anyhow, if ujson is faster for your work-flow we should use it. I assume you do not use the DBSCAN clustering?

**Comment by maxnoe** [comment]
uson spead up was a factor of 2 in doing `EventlistReader.inspect` on a 2700 events file.

The gzip people advise to use the magic bytes and not rely on file endings, it is far more reliable.

This is not a large PR. Netto line changes is well below 30 and the touched code is very clear and concise from my point of view.

**Pull request merged**